### PR TITLE
Add token auto-refreshing feature and support for DPS

### PR DESCRIPTION
**Pull request by ragrus-nbl**

### CODE REVIEW

File: AzureIoTHub.agent.lib.nut [hunk 1]
```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
...
+            const AZURE_DPS_GLOBAL_HOST = "https://global.azure-devices-provisioning.net";
+
+            // Looks like "https://global.azure-devices-provisioning.net/0ne000366B8/registrations/my-device/register?api-version=2018-09-01-preview"
+            const AZURE_DPS_REG_ENDPOINT_FMT        = "%s/%s/registrations/%s/register?api-version=%s";
+            // Looks like "https://global.azure-devices-provisioning.net/0ne000366B8/registrations/my-device/operations/4.176af959adcba99f.d5817a51-c64b-409b-83df-8425b610cbd2?api-version=2018-09-01-preview"
+            const AZURE_DPS_OP_STATUS_ENDPOINT_FMT  = "%s/%s/registrations/%s/operations/%s?api-version=%s";
+            // Looks like "https://global.azure-devices-provisioning.net/0ne000366B8/registrations/my-device?api-version=2018-09-01-preview"
+            const AZURE_DPS_REG_STATUS_ENDPOINT_FMT = "%s/%s/registrations/%s?api-version=%s";
+
+            const AZURE_DPS_SAS_TTL = 3600;
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: Should these two be configurable parameters?

File: AzureIoTHub.agent.lib.nut [hunk 2]
```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
...
+
+            request.sendasync(onSent);
+        }
+
+        // -------------------- PRIVATE METHODS -------------------- //
+
+        function _regOperationStatus(operationId, onCompleted) {
+            local url = format(AZURE_DPS_OP_STATUS_ENDPOINT_FMT, AZURE_DPS_GLOBAL_HOST, _scopeId, _regId, operationId, AZURE_DPS_API_VERSION);
+            local request = http.get(url, _headers);
+            local onSent = null;
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: No need to assign `null` here.

File: AzureIoTHub.agent.lib.nut [hunk 3]
```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
...
+                        onCompleted(AZURE_DPS_ERROR_NOT_REGISTERED, body, null);
+                        return;
+                    }
+                } else if (resp.statuscode != AZURE_DPS_TOO_MANY_REQS_CODE) {
+                    onCompleted(resp.statuscode, body, null);
+                    return;
+                }
+
+                if ("retry-after" in resp.headers) {
+                    retryAfter = resp.headers["retry-after"].tointeger();
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: Are these in seconds?

File: AzureIoTHub.agent.lib.nut [hunk 4]
```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
...
+        }
+
+        function _connectionString(iotHubAddr) {
+            return format("HostName=%s;DeviceId=%s;SharedAccessKey=%s", iotHubAddr, _regId, _deviceKey);
+        }
+
+        function _parseBody(body) {
+            try {
+                body = http.jsondecode(body);
+            } catch (e) {
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: no space between `catch (e)`

File: AzureIoTHub.agent.lib.nut [hunk 5]
```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
...
+            request.sendasync(onSent);
+        }
+
+        function _connectionString(iotHubAddr) {
+            return format("HostName=%s;DeviceId=%s;SharedAccessKey=%s", iotHubAddr, _regId, _deviceKey);
+        }
+
+        function _parseBody(body) {
+            try {
+                body = http.jsondecode(body);
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: I don't like the idea of modifying the argument even though it's a reference. I would introduce `local result` for that instead.

File: AzureIoTHub.agent.lib.nut [hunk 6]
```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
...
+                            onCompleted(0, body, _connectionString(body["assignedHub"]));
+                        } else {
+                            server.error("Response body doesn't have the \"assignedHub\" field");
+                            onCompleted(AZURE_ERROR_GENERAL, body, null);
+                        }
+                    } else {
+                        onCompleted(AZURE_DPS_ERROR_NOT_REGISTERED, body, null);
+                    }
+                } else if (resp.statuscode == AZURE_DPS_NOT_FOUND_CODE) {
+                    onCompleted(AZURE_DPS_ERROR_NOT_REGISTERED, body, null);
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: We might have discussed this already. If so, please remind me. 

Q1: we are replacing Azure error with our internal one. What's the point?
Q2: are we documenting `AZURE_DPS_ERROR_NOT_REGISTERED` error anywhere?

File: AzureIoTHub.agent.lib.nut (hunk 3 above)
Comment (ragrus-nbl) [reply AzureIoTHub.agent.lib.nut]: Yes

File: AzureIoTHub.agent.lib.nut (hunk 5 above)
Comment (ragrus-nbl) [reply AzureIoTHub.agent.lib.nut]: Fixed

File: AzureIoTHub.agent.lib.nut (hunk 2 above)
Comment (ragrus-nbl) [reply AzureIoTHub.agent.lib.nut]: Fixed

File: AzureIoTHub.agent.lib.nut [hunk 7]
```diff
@@ -732,44 +978,48 @@ class AzureIoTHub {
...
         // Closes the connection to Azure IoT Hub. Does nothing if the connection is already closed.
         //
         // Returns:                         Nothing.
         function disconnect() {
+            _log("Call: disconnect()");
             if ((!_isDisconnected || _isConnecting) && !_isDisconnecting) {
                 _isDisconnecting = true;
+                if (_isRefreshingToken) {
+                    _log("Token refreshing is in progress now. Putting the request (disconnect) to the queue...");
+                    _pendingCalls.append(@() _mqttclient.disconnect(_onDisconnected.bindenv(this)));
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: Don't we need to `bindenv(this)` the lambda function to let it access `_mqttclient `?

File: AzureIoTHub.agent.lib.nut [hunk 8]
```diff
@@ -794,15 +1045,23 @@ class AzureIoTHub {
...
-                onSent && onSent(_isConnected ? AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW : AZURE_IOT_CLIENT_ERROR_NOT_CONNECTED, msg);
+                onSent && onSent(_isConnected ? AZURE_CLIENT_ERROR_OP_NOT_ALLOWED_NOW : AZURE_CLIENT_ERROR_NOT_CONNECTED, msg);
                 return;
             }
 
-            local tooManyRequests = _msgPendingQueue.len() >= _options.maxPendingSendRequests;
+            if (_isRefreshingToken) {
+                _log("Token refreshing is in progress now. Putting the request (sendMessage) to the queue...");
+                // We don't do bindenv here because we do it for the function which processes the _pendingCalls queue
+                _pendingCalls.append(@() sendMessage(msg, onSent));
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: Same as above. Can you please double-check the rest of lambdas?

File: AzureIoTHub.agent.lib.nut [hunk 9]
```diff
@@ -850,17 +1110,25 @@ class AzureIoTHub {
...
             local disable = onReceive == null;
 
             if (!_readyToEnable(_isEnablingMsg, enabled, disable, onDone)) {
                 return;
             }
 
-            local doneCb = function (err) {
+            if (_isRefreshingToken) {
+                _log("Token refreshing is in progress now. Putting the request (enableIncomingMessages) to the queue...");
+                _pendingCalls.append(@() enableIncomingMessages(onReceive, onDone));
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: `bindenv(this)` missing?

File: AzureIoTHub.agent.lib.nut [hunk 10]
```diff
@@ -905,17 +1168,25 @@ class AzureIoTHub {
...
             local disable = onRequest == null;
 
             if (!_readyToEnable(_isEnablingTwin, enabled, disable, onDone)) {
                 return;
             }
 
-            local doneTwinRecvCb = function (err) {
+            if (_isRefreshingToken) {
+                _log("Token refreshing is in progress now. Putting the request (enableTwin) to the queue...");
+                _pendingCalls.append(@() enableTwin(onRequest, onDone));
```
Comment (ppetrosh) [inline AzureIoTHub.agent.lib.nut]: bindenv?

**Pull request merged**